### ISSUE TITLE
Fix `declaration-property-value-no-unknown` false negatives for typed custom properties

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#7061]
### What is the problem you're trying to solve?

The `@property` rule makes it possible to type custom properties.

```css
@property --foo {
	syntax: "<color>";
	inherits: false;
	initial-value: #c0ffee;
}
```

Custom properties registered in this way behave subtly different, for one they are animatable.  
But they are also type safe.

`declaration-property-value-no-unknown` doesn't validate the value types for these custom properties.

```css
.foo {
  --foo: bar; /* no warning */
}
```

### What solution would you like to see?

I would like to receive a warning when assigning values of an incorrect type to a typed custom property.

----

This depends on https://github.com/csstree/csstree/pull/256

----

The rule can be modified with this bit of code :

```js
/** @type {Record<string, string>} */
const discoveredProperties = {};

const typeSafeCustomProperties = new Set();

root.walkAtRules(/^property$/i, (atRule) => {
	const propName = atRule.params.trim();

	if (!propName) return;

	const syntax = atRule.nodes?.find((x) => isDeclaration(x) && /^syntax$/i.test(x.prop));

	if (!syntax || !isDeclaration(syntax)) return;

	typeSafeCustomProperties.add(propName);
	discoveredProperties[propName] = syntax.value;
});

const forkedLexer = fork({
	properties: {
		...propertiesSyntax,
		...discoveredProperties,
	},
	types: typesSyntax,
}).lexer;
```

**Comment by romainmenke** [comment]
Unsure how to best label this issue :)

- are these false negatives?
- is this a new sub-feature?
- should this be behind a plugin option?

**Comment by Mouvedia** [comment]
To me, it should a separate rule.
Sadly it would also be affected by #4088.

**Comment by romainmenke** [comment]
I hadn't considered this to be a separate rule because the subject (values of declarations) is the same and what is being checked (no unknown values) is also the same.

If it would be a separate rule it would also overlap with `declaration-property-value-no-unknown` if support is added for type checking `var()`.

```css
@property --foo {
	syntax: "<color>";
	inherits: false;
	initial-value: #c0ffee;
}

:root {
  --foo: red;
}

.component {
  width: var(--foo); /* I expect a warning */
}
```

I think this sub-feature would only add support for modern CSS syntax on the same subjects.

**Comment by Mouvedia** [comment]
> If it would be a separate rule it would also overlap with `declaration-property-value-no-unknown` **if support is added** for type checking `var()`.

As you know full well there are lot of PRs that are waiting to be merged on csstree side; it's unlikely this one will fare better.

> I hadn't considered this to be a separate rule…

It seems to me that it's more akin to type validation but you could be right: it might be on scope for `declaration-property-value-no-unknown`.

**Comment by jeddy3** [comment]
I think we can classify this as a false negative because this is something I should've considered when implementing the rule as it had [come up before](https://github.com/stylelint/stylelint/issues/5061#issuecomment-733642796).
